Various edits throughout the specification

jintekc opened the pull request:
Addresses the following issues
* #35 
* #58
* #98 
* #104
* #106 
* #114  
* #121
* #128
* #130
* #163

wip-abramson commented:
This looks great! Thanks for starting on this @jintekc 

Looks like you are also addressing #35

wip-abramson commented:
@jintekc can I ask that you keep your edits to this CRUD section

wip-abramson commented:
> @wip-abramson can you elaborate? what do you mean by that exactly?

I mean that if you make changes throughout the whole spec in a single PR we are going to run into challenges with merge conflicts. I would prefer to limit these.

jintekc commented on chapters/Index.md:
```diff
@@ -27,11 +27,14 @@ Kate Sills    <katelynsills@gmail.com>
+20th September 2024
+
+### Licence Statement: {.unnumbered .unlisted}
+TODO
```
will do!

jintekc commented on chapters/CRUD-Operations.md:
````diff
@@ -574,7 +642,7 @@ It returns the following output:
    ```
 1. Set `expectedProofPurpose` to `capabilityInvocation`.
-1. Set `mediaType` to ???? TODO: is this just `application/json`?
+1. Set `mediaType` to `application/ld+json`?
````
oops! good catch :smile: thanks @parasyte

jintekc commented on chapters/CRUD-Operations.md:
```diff
@@ -41,23 +44,27 @@ It returns the following outputs:
+   should include at least one verificationMethod and service of the type
+   [SingletonBeacon].
+* `version` - the identifier version, optional, defaults to 1.
+* `network` - the bitcoin network where the DID and DID Document can be resolved,
```
I think the recent move has been to lower case the "d" in "document" throughout the spec. Although I recall @dcdpr questioning this in our last call.

wip-abramson replied:
No, definitely lower case. We should copy the style that DID core uses here. As we are doing throughout the spec

```suggestion
* `network` - the bitcoin network where the DID and DID document can be resolved,
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -41,23 +44,27 @@ It returns the following outputs:
-to create a **did:btc1** from some intermediate DID document. This allows for a
+to create a **did:btc1** from some external intermediate DID document. This allows for a
 more complex ::Initial DID Document::, including the ability to include
 Service Endpoints and ::Beacons:: that support aggregation.
```
```suggestion
Service Endpoints and ::Beacons:: that support aggregation.
```

@jenniemeier should this be capitalized? Not sure where we landed

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -81,26 +88,30 @@ It returns the following outputs:
 executed by a resolver after a resolution request identifying a specific
 **did:btc1** `identifier` is received from a client at ::Resolution Time::.
-The request MUST always contain the resolutionOptions object containing additional
+The request MUST always contain the `resolutionOptions` object containing additional
 information to be used in resolution. This object MAY be empty. 
```
```suggestion
The request MUST always contain the `resolutionOptions` object containing additional
information to be used in resolution. This object MAY be empty. 
See the [DID Resolution specification](https://w3c.github.io/did-resolution/#did-resolution-options) for further details about the DID Resolution Options object.
```

I am sure there are many places throughout the spec that could benefit from better referencing. This seems like an easy one to catch now

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -81,26 +88,30 @@ It returns the following outputs:
-   * `versionTime`: a timestamp used during resolution as a bound for when to
+
+* `identifier` - a valid **did:btc1** identifier
+* `resolutionOptions` - an optional object containing the following options
```
```suggestion
* `resolutionOptions` - an object containing the following optional options
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -113,20 +124,29 @@ It returns the following output:
 
-The Resolve ::Initial DID Document:: algorithm specifies how to resolve an ::Initial
-DID Document:: and validate it against the `identifier` for a specific **did:btc1**.
+This algorithm specifies how to resolve an initial DID document and validate
```
```suggestion
This algorithm specifies how to resolve an ::initial DID document:: and validate
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -113,20 +124,29 @@ It returns the following output:
+* `identifierComponents` - An object containing the following properties
+  * `idType` - the type of identifier (`KEY` or `EXTERNAL`)
+  * `version` - the identifier version
+  * `network` - the identifier bitcoin network
```
```suggestion
  * `network` - the identifier Bitcoin network
```

I believe

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -113,20 +124,29 @@ It returns the following output:
+  * `version` - the identifier version
+  * `network` - the identifier bitcoin network
+  * `genesisBytes` - the originating public key bytes
+* `resolutionsOptions` - optional object that extends the default
+   [DID Resolution Options](https://www.w3.org/TR/did-1.0/#did-resolution-options)
+  * `versionId` - TODO: What exactly is this?
+  * `versionTime` - the time at which the identifier and DID document were created
+  * `sidecarData` - data necessary for resolving a DID such as ::DID Update Payloads:: and ::SMT proofs::.
+  * `network` - the identifier bitcoin network
```
This is nicer than above. We should use this style both times I feel.

Also this TODO for versionId. versionId in did:btc1 is an incrementing integer starting from 1, the version of the initial document

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -144,15 +164,19 @@ The Deterministically Generate ::Initial DID document:: algorithm generates an
+* `identifierComponents` - An object containing the following properties
+  * `idType` - the type of identifier (KEY or EXTERNAL)
+  * `version` - the identifier version
+  * `network` - the identifier bitcoin network
```
```suggestion
  * `network` - the identifier Bitcoin network
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -185,25 +209,36 @@ It returns the following output:
-   * `type`: a string signaling the type of beacon, default='SingletonBeacon'
-   * `id`: the **did:btc1** identifier controlling the beacon plus the DID fragment pointing
+
+* `services` - an array of beacon services objects containing the following properties
```
```suggestion
* `services` - an array of BTC1 Beacon services objects containing the following properties
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -234,16 +269,20 @@ as an `intermediateDocumentRepresentation`, either by retrieving it from
+* `identifierComponents` - An object containing the following properties
+  * `idType` - the type of identifier (KEY or EXTERNAL)
+  * `version` - the identifier version
+  * `network` - the identifier bitcoin network
```
```suggestion
  * `network` - the identifier Bitcoin network
```

Could you just check the whole PR for this catch @jintekc

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -234,16 +269,20 @@ as an `intermediateDocumentRepresentation`, either by retrieving it from
+  * `version` - the identifier version
+  * `network` - the identifier bitcoin network
+  * `genesisBytes` - the originating public key bytes
+* `resolutionOptions` - A set of optional parameters used for resolution
```
Should use style I referred to above I feel

parasyte commented on chapters/CRUD-Operations.md:
```diff
@@ -41,23 +44,27 @@ It returns the following outputs:
-   ::SingletonBeacon::.
-* `version`: the identifier version, optional, defaults to 1.
-* `network`: the bitcoin network where the DID and DID document can be resolved,
+   should include at least one verificationMethod and service of the type
```
I noticed that this change (to lowercase) made the line consistent with other uses of "should" within the spec. Is the recommendation to capitalize all normative uses, or just this one?

jintekc replied:
Yeah, I think I normalize to the wrong case. All should be normalized to CAPS. I'll add that to the PR.

jintekc commented:
I've applied all suggestions and added some more edits for alignment such as:
* DID Document => DID document (except when found in a header)
* Aligning the definition of arguments across algorithms where appropriate
* Added types and REQUIRED/OPTIONAL to algorithm inputs
* Linked key terms and external content

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -40,24 +43,28 @@ It returns the following outputs:
-* `network`: the bitcoin network where the DID and DID document can be resolved,
-   optional, defaults to `bitcoin`.
+   SHOULD include at least one `verificationMethod` and `service` of the type
+   [SingletonBeacon]; REQUIRED; object.
```
What does the [ ] do?

jintekc replied:
I think I meant to do ::SingletonBeacon::

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -138,21 +165,25 @@ It returns the following output:
 It returns the following output:
-* `initialDocument`: a valid DID document for the given identifier.
+
+* `initialDocument` - the valid first version of a DID document for a given btc1 identifier.
```
```suggestion
* `initialDocument` - the valid first version of a DID document for a given *did:btc1*** identifier.
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -185,25 +216,37 @@ It returns the following output:
+
+Spends from these three addresses can be produced only through signatures from the
+private key associated with the `keyBytes`. Each ::Beacon:: is of the type
+[SingletonBeacon].
```
I wonder if this should be Each BTC1 Beacon is a Singleton Beacon. With the terms referenced. This goes to the singleton beacon section does it?

jintekc replied:
Sounds good. Yeah, it should have :: around it. Will fix.

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -228,22 +271,35 @@ It returns the following output:
+The External Resolution algorithm externally retrieves an ::intermediate DID document::
 as an `intermediateDocumentRepresentation`, either by retrieving it from
-::Content Addressable Storage:: (CAS) or from the ::Sidecar Data:: provided as
+::ContentiAddressable Storage:: (CAS) or from the ::Sidecar Data:: provided as
```
```suggestion
::Content Addressable Storage:: (CAS) or from the ::Sidecar Data:: provided as
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -317,15 +381,27 @@ announce ::DID Update Payloads:: applicable to the **did:btc1** identifier being
+   A DID Core conformant DID document; REQUIRED; object
+* `resolutionOptions` - options that extends the default options per
+   the [DID Resolution specification](https://w3c.github.io/did-resolution/#did-resolution-options);
+   OPTIONAL; object
```
This this is okay. Just still wondering how to indicate that this set of options below is not intended to be exhaustive

jintekc replied:
I'll add a note

```

It takes the following inputs:

* `identifier` - a valid **did:btc1** identifier; REQUIRED; string
* `resolutionOptions` - an object that extends the default options per
   the [DID Resolution specification](https://w3c.github.io/did-resolution/#did-resolution-options);
   the below list is not intended to be exhaustive; OPTIONAL; object
   * `versionId` - the version of the DID and DID document, an incrementing integer starting
      from 1; OPTIONAL; integer
   * `versionTime` - a timestamp used during resolution as a bound for when to
      stop resolving; OPTIONAL; integer
   * `sidecarData` - data necessary for resolving a DID such as ::BTC1 Updates:: and
      ::SMT proofs::; OPTIONAL; object.
   * `network` - the Bitcoin network used for resolution; OPTIONAL; string; default=`"bitcoin"`
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -358,47 +434,50 @@ reached, or the blockchain history passes the supplied `targetTime`. At this
-* `targetTime`: A 64-bit UNIX timestamp that can be used to target specific historical states of a DID document. 
+   blockchain history. an integer greater of equal to 0; REQUIRED; integer
+* `currentVersionId` - the version of the ::contemporary DID document::. An integer starting from 
+   1 and incrementing by 1 with each ::DID Update Payload:: applied to the DID document; REQUIRED; integer
```
payload to update pls

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -358,47 +434,50 @@ reached, or the blockchain history passes the supplied `targetTime`. At this
-      the value at the leaf indexed by the **did:btc1** being resolved. TODO: What exactly this
+      DID Update Payload MAY be provided; OPTIONAL; object
+   * `proofs` - a ::Sparse Merkle Tree:: proof that the provided `updatePayload` value is
+      the value at the leaf indexed by the **did:btc1** being resolved. REQUIRED; TODO: What exactly this
```
@KDean-Dolphin can you chime in here. This is defined somewhere now (in your pending PR.)  So happy to get this in as is. But we need to track that this TODO needs fixing asap

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -358,47 +434,50 @@ reached, or the blockchain history passes the supplied `targetTime`. At this
-1. Find all `beacons` in `contemporaryDIDDocument`: All `service` in
-   `contemporaryDIDDocument.service` where `service.type` equals one of
-   `SingletonBeacon`, `CIDAggregateBeacon` and `SMTAggregateBeacon` ::Beacon::.
+   [JSON Canonicalization and Hash] algorithm.
```
```suggestion
   [JSON Canonicalization and Hash] algorithm.
```

Wondering how we handle Ryans desire to move to RDFC. I am inclined to put a flag at the front of the spec. But leave the spec as is for now.

I do not think we are going to be able to get our libraries in sync and aligned using RDFC in time. Wheras today, they are in sync and working. I do not see a security concern with this personally. It is clearly a distinct crypto suite. And these are just going on the test network anyway. We are not version 1 yet. This is one of the areas we expect to change

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -486,34 +571,37 @@ It returns the following output:
 
-The Process Beacon Signals algorithm processes each ::Beacon Signal:: by attempting to retrieve and validate an announce 
-::DID Update Payload:: for that signal according to the type of the ::Beacon::.
+The Process Beacon Signals algorithm processes each ::Beacon Signal:: by attempting to retrieve and
+validate an announce ::DID Update Payload:: for that signal according to the type of the ::Beacon::.
```
```suggestion
The Process Beacon Signals algorithm processes each ::Beacon Signal:: by attempting to retrieve a ::BTC1 Beacon Announcement:: for a specific **did:btc1** identifier.
Then the algorithm retrieves and validates the ::BTC1 Update:: committed to by the announcement. 
The ::Beacon Type:: of the ::BTC1 Beacon:: that broadcast the ::Beacon Signal:: defines the algorithms to retrieve and validate 
::BTC1 Beacon Announcements:: and their associated ::BTC1 Updates"" against the ::Beacon Signal::.
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -486,34 +571,37 @@ It returns the following output:
 It returns the following output:
-* `updates`: an array of ::DID Update Payloads::
+
+* `updates` - an array of ::DID Update Payloads::
```
```suggestion
* `btc1Updates` - an array of ::BTC1 Updates::
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -595,35 +690,41 @@ It returns the following output:
+   REQUIRED; string.
+* `beaconIds` - an array containing the IDs that correspond to ::Beacons:: in the DID
+   document. These MUST identify service endpoints with one of the three ::Beacon Types::
+   `SingletonBeacon`, `CIDAggregateBeacon`, and `SMTAggregateBeacon`; REQUIRED; array.
```
```suggestion
   `SingletonBeacon`, `MapBeacon`, and `SMTBeacon`; REQUIRED; array.
```

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -669,31 +777,33 @@ It returns the following output:
-   use for ZCAP-LD
+
+* `identifier` - a valid **did:btc1** identifier; REQUIRED; string.
+* `didUpdatePayload` - an unsigned ::DID Update Payload::; REQUIRED; object.
```
payload to btc1Update

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -726,22 +833,27 @@ and calls the Broadcast DID Update algorithm corresponding to the type of
+* `beaconIds` - an array containing the IDs that correspond to ::Beacons:: in the DID
+   document. These MUST identify service endpoints with one of the three ::Beacon Types::
+   `SingletonBeacon`, `CIDAggregateBeacon`, and `SMTAggregateBeacon`; REQUIRED; array.
+* `didUpdateInvocation` - a ::DID Update Payload:: object invoked and secured with
```
This would require a pass to make sure we are referring to the right things in the right places

jintekc replied:
Sounds good

wip-abramson commented on chapters/CRUD-Operations.md:
```diff
@@ -726,22 +833,27 @@ and calls the Broadcast DID Update algorithm corresponding to the type of
+* `signalsMetadata` - a mapping from Bitcoin transaction identifiers of ::Beacon Signals:: 
+   to a struct containing ::Sidecar Data:: for that signal, provided as part of the
+   `resolutionOptions`; a map with the following properties:
+   * `updatePayload` - a ::DID Update Payload:: which SHOULD match the update announced
```
payload to btc1Update

jintekc commented on chapters/CRUD-Operations.md:
```diff
@@ -411,8 +490,8 @@ It returns the following output once either `targetTime` or `targetVersionId` ha
-       run the [Confirm Duplicate Update] Algorithm passing in `update`,
-       `updateHashHistory`, and `contemporaryHash`.
+       push `contemporaryHash` into the `updateHashHistory` list and run the
+       [Confirm Duplicate Update] algorithm passing in `update` and `updateHashHistory`.
```
I think it makes it overly explicit, which is our goal, but idk that it adds much value since its just var naming. Happy to make the change throughout though to remain consistent.

wip-abramson replied:
Fair enough. Well I definitely want to change anywhere we use `updatePayload`. That should either be `update` or `btc1Update`. I feel for 4 chars we get more clarity which is worthwhile and it maps to our terms

wip-abramson commented on chapters/CRUD-Operations.md:
````diff
@@ -185,25 +216,37 @@ It returns the following output:
-   ```{.json include="json/CRUD-Operations/Read-deterministically-generate-beacon-services.json"}
-   ```
+
+* `services` - an array of ::BTC1 Beacon:: service objects containing the following properties
+   * `type` - described the kind of service being defined used to determine how to produce updates; 
+      REQUIRED; string; default=`"SingletonBeacon"`
+   * `id` - the **did:btc1** identifier controlling the beacon including the DID fragment pointing
+      to the location of a key in the DID document; REQUIRED; string
+   * `serviceEndpoint` - a P2PKH, P2WPKH or P2TR bitcoin address where beacon signals can be broadcast; REQUIRED; string
````
I think this is fine in this PR, but just want to flag that we will need to update this later regarding the new, still slightly debated, type structure for Beacons. i.e. type: ["BTC1Beacon", "SingletonBeacon"]. Or the other variation

wip-abramson commented:
Okay! I can take a pass RE that. A DID Update Invocation IS a BTC1 Update now. A BTC1 Update is an Unsecured BTC1 Update with a capability invocation Data INtegrity proof attached to it.

That proof invokes the capability to update a specific DID document.
